AlphaWallet/KeyManagement/SecureEnclave.swift

```diff
@@ -0,0 +1,176 @@
+                return "Encryption key already exist for: \(name)"
+            case .cannotAccessPrivateKey(let osStatus):
+                return "Cannot access private key because: \(osStatus)"
+            case .cannotAccessPublicKey:
+                return "Cannot access public key"
+            case .encryptionNotSupported(let algorithm):
+                return "Encryption not supported: \(algorithm)"
+            case .decryptionNotSupported(let algorithm):
+                return "Decryption not supported: \(algorithm)"
+            case .cannotEncrypt:
+                return "Cannot encrypt"
+            case .cannotDecrypt:
+                return "Cannot decrypt"
+            case .unexpected(let description):
+                return description
+            }
+        }
+    }
+
+    private let requiresBiometry = true
+    private let numberOfBitsInKey = 256
+    private let algorithm = SecKeyAlgorithm.eciesEncryptionCofactorVariableIVX963SHA256AESGCM
+    private let userPresenceRequired: Bool
+
+    private var isSimulator: Bool {
+        return TARGET_OS_SIMULATOR != 0
+    }
+
+    init(userPresenceRequired: Bool = false) {
+        self.userPresenceRequired = userPresenceRequired
+    }
+
+    private func getPrivateKey(withName name: String) throws -> SecKey {
+        let params: [String: Any] = [
+            kSecClass as String: kSecClassKey,
+            kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
+            kSecAttrApplicationTag as String: tagData(fromName: name),
+            kSecReturnRef as String: true,
+        ]
+
+        var raw: CFTypeRef?
+        let status = SecItemCopyMatching(params as CFDictionary, &raw)
+        guard status == errSecSuccess, let result = raw else {
+            throw Error.cannotAccessPrivateKey(osStatus: status)
+        }
+        return result as! SecKey
+    }
+
+    private func getPrivateKeyCount(withName name: String) -> Int {
+        let params: [String: Any] = [
+            kSecClass as String: kSecClassKey,
+            kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
+            kSecAttrApplicationTag as String: tagData(fromName: name),
+            kSecReturnRef as String: true,
+            kSecMatchLimit as String: kSecMatchLimitAll
+        ]
+
+        var raw: CFTypeRef?
+        let status = SecItemCopyMatching(params as CFDictionary, &raw)
+        if status == errSecSuccess, let all = raw as? [SecKey] {
+            return all.count
+        } else {
+            return 0
+        }
+    }
+
+    private func encrypt(plainTextData: Data, withPublicKey publicKey: SecKey) throws -> Data {
+        guard SecKeyIsAlgorithmSupported(publicKey, .encrypt, algorithm) else {
+            throw Error.encryptionNotSupported(algorithm: algorithm)
+        }
+        var error: Unmanaged<CFError>?
+        guard let cipherTextData = SecKeyCreateEncryptedData(publicKey, algorithm, plainTextData as CFData, &error) as? Data else {
```

Comment:
What happens if the seed is larger than the blocksize with which we are encrypt using ECDSA? Is an error thrown or should this be checked explicitly before the call? Some comment goes for `decrypt` call.

Reply:
The docs says that `SecKeyCreateEncryptedData` and `SecKeyCreateDecryptedData` will return nil upon error. We check that and throw a Swift exception (which we then handle in the UI client code) when it fails. But I tested encrypting a long plaintext and decrypting the result and at least it works empirically.